multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/BuildCloudDeployModelStep.java

```diff
@@ -148,6 +161,110 @@ private List<List<Resource>> getResourceBatches(ProcessContext context, List<Res
+
+            List<String> newManagedKeys = newKeyNamesByService.getOrDefault(serviceName, Collections.emptyList());
+            List<String> existingKeys = existingKeysByService.getOrDefault(serviceName, Collections.emptyList());
+
+            List<DeployedMtaServiceKey> unusedKeys = deployedKeys.getValue()
+                                                                 .stream()
+                                                                 .filter(deployedKey -> keyIsMissingInBothLists(deployedKey.getName(),
+                                                                                                                newManagedKeys,
+                                                                                                                existingKeys))
+                                                                 .collect(toList());
+
+            keysToDelete.addAll(unusedKeys);
+        }
+
+        getStepLogger().info("DEBUG: keysToDelete \"{0}\"", keysToDelete);
+        return keysToDelete;
+
+    }
+
+    private boolean keyIsMissingInBothLists(String keyName, List<String> firstList, List<String> secondList) {
+        return !(firstList.contains(keyName) || secondList.contains(keyName));
+    }
+
+    private Map<String, List<String>> mapKeysByServiceName(Map<String, List<CloudServiceKey>> keysByResource,
+                                                           Map<String, List<CloudServiceKey>> additionalKeysByResource) {
+        Map<String, List<String>> keyNamesMap = new HashMap<>();
+
+        addKeysNamesToMap(keyNamesMap, keysByResource);
+        addKeysNamesToMap(keyNamesMap, additionalKeysByResource);
+
+        return keyNamesMap;
+    }
+
+    private void addKeysNamesToMap(Map<String, List<String>> allKeysMap, Map<String, List<CloudServiceKey>> keysByResource) {
+        if (keysByResource == null || keysByResource.isEmpty()) {
+            return;
+        }
+
+        for (List<CloudServiceKey> keysForResource : keysByResource.values()) {
+            if (!CollectionUtils.isEmpty(keysForResource)) {
+                addServiceKeysToMap(keysForResource, allKeysMap);
+            }
+        }
+    }
+
+    private void addServiceKeysToMap(List<CloudServiceKey> keysForResource, Map<String, List<String>> allKeysMappedByService) {
+        String serviceName = keysForResource.get(0)
+                                            .getServiceInstance()
+                                            .getName();
+
+        List<String> keyNames = keysForResource.stream()
+                                               .map(key -> key.getName())
+                                               .collect(toList());
+
+        if (allKeysMappedByService.containsKey(serviceName)) {
+            allKeysMappedByService.put(serviceName, ListUtils.union(allKeysMappedByService.get(serviceName), keyNames));
+        } else {
+            allKeysMappedByService.put(serviceName, keyNames);
+        }
```

Comment:
This can be replaced with:
```java
allKeysMappedByService.computeIfPresent(serviceName, (sName, currentKeys) -> ListUtils.union(currentKeys, keyNames));
```

Reply:
also replaced with merge